utils/general.py

```diff
@@ -0,0 +1,183 @@
+import os
+from typing import List, Tuple
+
+import numpy as np
+
+import torch
+import torch.nn as nn
+import transformers
+
+from abnumber import Chain
+
+import antiberty
+from antiberty import AntiBERTy
+
+from glob import glob
+import igfold
+
+from igfold.model.IgFold import IgFold
+from igfold.utils.folding import fold
+from igfold.utils.embed import embed
+from igfold.utils.general import exists
+
+from igfold.model.interface import IgFoldInput
+from igfold.utils.fasta import get_fasta_chain_dict
+from igfold.utils.general import exists
+from igfold.utils.pdb import get_atom_coords, save_PDB, write_pdb_bfactor, cdr_indices
+
+project_path = os.path.dirname(os.path.realpath(antiberty.__file__))
+trained_models_dir = os.path.join(project_path, 'trained_models')
+ANTIBERTY_CHECKPOINT_PATH = os.path.join(trained_models_dir, 'AntiBERTy_md_smooth')
+ANTIBERTY_VOCAB_FILE = os.path.join(trained_models_dir, 'vocab.txt')
+
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+def exists(x):
+    return x is not None
+
+def freeze(model):
+    for p in model.parameters():
+        p.requires_grad = False
+
+class FVbinder:
+    def __init__(self,
+                 vh: str, 
+                 vl: str, 
+                 coords: torch.Tensor = None, 
+                 vh_cdr_mask: torch.Tensor = None, 
+                 vl_cdr_mask: torch.Tensor = None
+                ):
+        self.vh = vh
+        self.vl = vl
+        self.coords = coords
+        self.vh_cdr_mask = vh_cdr_mask
+        if self.vh_cdr_mask is None:
+            self.vh_cdr_mask = self.get_mask_from_sequence(self.vh).to(device)
+        self.vl_cdr_mask = vl_cdr_mask
+        if self.vl_cdr_mask is None:
+            self.vl_cdr_mask = self.get_mask_from_sequence(self.vl).to(device)
+
+    @staticmethod
+    def get_mask_from_sequence(sequence: str) -> torch.Tensor:
+        mask = torch.zeros((len(sequence), 5))
+        a = Chain(sequence, scheme="chothia")
+        for reg, positions in a.regions.items():
+            if reg.startswith("FR"):
+                continue
+            for pos, amino in positions.items():
+                mask[pos.number] = torch.ones(5)
+        return mask
+
+    def to_pdb(self, filename: str):
+        assert self.coords is not None
```

Comment:
assert-ов лучше избегать в зрелом коде, потому что они не дают понимания произошедшей ошибки.
```suggestion
        if self.coords is None:
            raise ValueError('no coordinates data')
```